omnigibson/transition_rules.py

```diff
@@ -369,213 +517,192 @@ def transition(self, individual_objects, group_objects):
+            if "cooked" in name or not fillable_obj.states[ContainsAny].get_value():
+                continue
+            # TODO: Remove this assert once we have a more standardized method of globally R/W particle positions
+            assert len(system.particle_instancers) == 1, \
+                f"PhysicalParticleSystem {system.name} should only have one instancer!"
+            # Replace all particles inside the container with their cooked versions
+            cooked_system = get_system(f"cooked_{system.name}")
+            positions = contains_state.cache[(system,)]["info"]["positions"]
+            in_volume = contains_state.cache[(system,)]["info"]["in_volume"]
+            in_volume_idx = np.where(in_volume)[0]
+            system.default_particle_instancer.remove_particles(idxs=in_volume_idx)
+            cooked_system.default_particle_instancer.add_particles(positions=positions[in_volume_idx])
 
-        # Create a new object to be added.
-        all_pos, all_orn = [], []
-        for contained_obj in contained_objs:
-            pos, orn = contained_obj.get_position_orientation()
-            all_pos.append(pos)
-            all_orn.append(orn)
-
-        category = self.obj_attrs.category
-        model = self.obj_attrs.model
-        name = f"{self.obj_attrs.name}_{self._counter}"
-        self._counter += 1
-        scale = self.obj_attrs.scale
-
-        model_root_path = f"{gm.DATASET_PATH}/objects/{category}/{model}"
-        usd_path = f"{model_root_path}/usd/{model}.usd"
-
-        final_obj = DatasetObject(
-            prim_path=f"/World/{name}",
-            usd_path=usd_path,
-            category=category,
-            name=f"{name}",
-            scale=scale)
-
-        final_obj_attrs = ObjectAttrs(
-            obj=final_obj, pos=np.mean(all_pos, axis=0), orn=np.mean(all_orn, axis=0))
-        t_results.add.append(final_obj_attrs)
-
-        # Delete all objects inside the container.
-        for contained_obj in contained_objs:
-            t_results.remove.append(contained_obj)
-
-        # Turn off the container, otherwise things would turn into garbage.
-        if ToggledOn in container_obj.states:
-            container_obj.states[ToggledOn].set_value(False)
-        print(f"Applied {ContainerRule.__name__} to {container_obj}")
         return t_results
 
 
-class ContainerGarbageRule(BaseTransitionRule):
+class MeltingRule(BaseTransitionRule):
     """
-    Rule to apply to a container to turn what remain inside into garbage.
-
-    This rule is used as a catch-all rule for containers to turn objects inside
-    the container that did not match any other legitimate rules all into a
-    single garbage object.
+    Transition rule to apply to meltable objects to simulate melting
     """
+    @classproperty
+    def individual_filters(cls):
+        # We want to find all meltable object / heatsource combos
+        # TODO: This does not handle flammable objects currently
```

Comment:
originally, i thought that it removes the assumption that all filters are static (i.e.: unchanging over time given the same set of objects).

Though actually, now that I think of it, I think this is actually fine, since the number of objects that have the OnFire state is static, and we'll just check for OnFire == True. Updated!